### PR TITLE
Adds ability to generate and send InvoicePDF as an attachment.

### PR DESCRIPTION
**Pull request by mohd-anas-ansari**
## Notion card
https://www.notion.so/saeloun/Generate-PDF-attachment-of-the-invoice-21035ae1b67a4aa7820b1ae21d7290b6
## Summary
Added generation of PDF AS attachment while mailing invoices.

## Preview

Screenshot 2022-04-28 at 2 46 04 AM

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

### Checklist:

- [x] I have manually tested all workflows
- [x] I have performed a self-review of my own code
- [ ] I have added automated tests for my code

### CODE REVIEW

**Comment by apoorv1316** [comment]
@anasansari157 CI check failed.

File: package.json [hunk 1]
```diff
@@ -33,6 +33,7 @@
     "phosphor-react": "^1.4.1",
     "postcss": "~7",
     "prop-types": "^15.7.2",
+    "puppeteer": "^13.6.0",
```
Comment (akhilgkrishnan) [inline package.json:36]: why do we need puppeteer

File: app/mailers/invoice_mailer.rb [hunk 2]
```diff
@@ -9,6 +11,21 @@ def invoice
...
+        company_logo: @invoice.client.company.logo.attached? ? polymorphic_url(@invoice.client.company.logo) : ""
+      })
+
+    style_tag_options = [
+      { content: "h1{background-color: red}" }
+    ]
+    pdf = Grover.new(html, style_tag_options:).to_pdf
+    attachments["invoice_#{@invoice.id}.pdf"] = pdf
+
     mail(to: recipients, subject:)
```
Comment (akhilgkrishnan) [inline app/mailers/invoice_mailer.rb:29]: add a reply to, with company email id

**Review by akhilgkrishnan — COMMENTED** [review]
@anasansari157  Added some suggestions, also Can you add test for helpers and mailers.

File: package.json (hunk 1 above)
Comment (anasansari157) [reply package.json:36]: It is a dependency for Grover`. We are using `Grover` for PDFs

File: app/views/invoices/pdf.html.erb [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+  <div class="px-10 pt-5">
+    <table class="w-full table-fixed">
+      <thead class="my-2">
+        <tr>
+          <th class="text-miru-dark-purple-600 font-normal text-xs text-left tracking-widest">NAME</th>
+          <th class="px-3 text-miru-dark-purple-600 font-normal text-xs text-left tracking-widest">DATE</th>
+          <th class="pl-2 text-miru-dark-purple-600 font-normal text-xs text-left tracking-widest w-1/3">DESCRIPTION</th>
+          <th class="text-miru-dark-purple-600 font-normal text-xs text-right tracking-widest">RATE</th>
+          <th class="text-miru-dark-purple-600 font-normal text-xs text-right tracking-widest">QTY</th>
+          <th class="text-miru-dark-purple-600 font-normal text-xs text-right tracking-widest">L. TOTAL</th>
```
Comment (akhilgkrishnan) [inline app/views/invoices/pdf.html.erb:77]: typo?

File: app/mailers/invoice_mailer.rb (hunk 2 above)
Comment (anasansari157) [reply app/mailers/invoice_mailer.rb:30]: We don't have email for companies.

Comment (akhilgkrishnan) [reply app/mailers/invoice_mailer.rb:30]: We have

Comment (akhilgkrishnan) [reply app/mailers/invoice_mailer.rb:30]: Ohh yes, we don't have. Can we add no-reply email id

File: app/views/invoices/pdf.html.erb [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        <tr>
+          <td class="pt-4 font-normal text-base text-miru-dark-purple-1000 text-right pr-10">Tax</td>
+          <td class="pt-4 font-bold text-base text-miru-dark-purple-1000 text-right w-22">
+            <%="#{invoice.client.company.base_currency} #{invoice["tax"]}"%>
+          </td>
+        </tr>
+        <tr>
+          <td class="pt-1 font-normal text-base text-miru-dark-purple-1000 text-right pr-10">Total</td>
+          <td class="font-bold text-base text-miru-dark-purple-1000 text-right">
+            <%="#{invoice.client.company.base_currency} #{(sub_total + invoice["tax"].to_i - invoice["discount"].to_i)}"%>
```
Comment (rohitjoshixyz) [inline app/views/invoices/pdf.html.erb:123]: Extract this calculation into a private method in the mailer

**Comment by apoorv1316** [comment]
@mohd-anas-ansari Have you fixed all the review comments on this PR?

**Comment by mohd-anas-ansari** [comment]
@apoorv1316 Yes, all reviews have been addressed.

File: app/helpers/invoices_helper.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+module InvoicesHelper
+  def serializer(invoice_line_items)
```
Comment (rohitjoshixyz) [inline app/helpers/invoices_helper.rb:4]: Why is this method called serializer? It is returning a hash if I am not mistaken

File: app/mailers/invoice_mailer.rb [hunk 6]
```diff
@@ -9,6 +11,27 @@ def invoice
...
+
+    controller = ActionController::Base.new
+    html = controller.render_to_string(
+      template: "invoices/pdf",
+      layout: "layouts/pdf",
+      locals: {
+        invoice: @invoice,
+        company_logo: @invoice.client.company.logo.attached? ?
+                        polymorphic_url(@invoice.client.company.logo) :
+                        "",
```
Comment (rohitjoshixyz) [inline app/mailers/invoice_mailer.rb:24]: Move to a private method, difficult to read here

File: app/mailers/invoice_mailer.rb [hunk 7]
```diff
@@ -9,6 +11,27 @@ def invoice
...
-    mail(to: recipients, subject:)
+    serialized_invoice = serializer(@invoice.invoice_line_items)
+
+    controller = ActionController::Base.new
+    html = controller.render_to_string(
+      template: "invoices/pdf",
+      layout: "layouts/pdf",
+      locals: {
+        invoice: @invoice,
+        company_logo: @invoice.client.company.logo.attached? ?
```
Comment (rohitjoshixyz) [inline app/mailers/invoice_mailer.rb:22]: You have added association for invoice and company in this same PR, please use that `@invoice.company`

**Review by rohitjoshixyz — APPROVED** [review]
Some non-blocking comments, LGTM 🎉

File: app/mailers/invoice_mailer.rb (hunk 7 above)
Comment (shivamsinghchahar) [inline app/mailers/invoice_mailer.rb:22]: Invoice generation is not the responsibility of the mailer. We should have a separate service for this.

**Review by shivamsinghchahar — CHANGES_REQUESTED** [review]

**Comment by supriya3105** [comment]
@apoorv1316 Please test this .

**Comment by apoorv1316** [comment]
@shivamsinghchahar Please approve it so it can be merged.

**Review by shivamsinghchahar — APPROVED** [review]
LGTM🚀

**Pull request merged**